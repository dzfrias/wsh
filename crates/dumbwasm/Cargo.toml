[package]
name = "dumbwasm"
version = "0.1.0"
edition = "2021"

[lints]
workspace = true

[dependencies]
wasm-leb128 = { path = "../wasm-leb128" }
logos = "0.13.0"
snailquote = "0.3.1"
anyhow = "1.0.75"
byteorder = "1.5.0"
clap = { version = "4.4.6", features = ["derive"] }
thiserror = "1.0.49"
console = "0.15.7"
