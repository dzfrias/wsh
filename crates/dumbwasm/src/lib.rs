//! # dumbwasm
//!
//! dumbwasm is an extremely simple text-to-[WebAssembly](https://webassembly.org/) transpiler.
//! Unlike the existing WebAssembly text format (WAT), dumbwasm syntax allows for invalid modules
//! to be generated. This makes dumbwasm ideal for writing tests related to WebAssembly!
//!
//! Note that if you're just looking to compile text to WebAssembly, I'd recommend looking using
//! the standard text format. dumbwasm is intended for people who need to manipulate the binary
//! format in ways that WAT doesn't allow.
//!
//! ## Syntax
//!
//! Here is a super simple WebAssembly module, in dumbwasm!
//!
//! ```dumbasm
//! magic
//! version
//! type {
//!   (anything in parens is skipped by the transpiler)
//!   [
//!     form.func 1 (params) i32 1 (result) i32,
//!   ]
//! }
//! function {
//!   ;; Equivalent to writing `[0]`! The square brackets automatically count up
//!   ;; the number of items!
//!   1 (number of funcs)
//!   0 (index)
//! }
//! code {
//!   [
//!     ;; This function returns 10 + 20
//!     instrs {
//!       [] (number of locals)
//!       i32.const 10
//!       ;; Base 16 also supported!
//!       i32.const 0xfafa
//!       i32.add
//!       return
//!     },
//!   ]
//! }
//! ```
//!
//! ### Sections
//!
//! You can create a section using the appropriate section name, along with curly braces. This will
//! automatically tally up section size for you, so you don't have to count up all the bytes by
//! hand!
//!
//! Here are the section names:
//!
//! - `custom`
//! - `type`
//! - `import`
//! - `function`
//! - `table`
//! - `memory`
//! - `global`
//! - `export`
//! - `start`
//! - `element`
//! - `code`
//! - `data`
//! - `datacount`
//!
//! ### Instructions
//!
//! Instructions will directly be translated into their binary form. For example, the `unreachable`
//! instruction translates to `0x00`. Instructions can be used anywhere in the module.
//!
//! There's also the `instrs` construct, which (like sections) automatically inserts the size into
//! the binary, to save you from having to count up by hand. It is important to note that `instrs`
//! only does the counting, no more and no less. Instructions are valid anywhere. This is key for
//! things like const initialization expressions, which don't have a size component in the binary
//! format.
//!
//! ### Literals
//!
//! Literals of many forms are allowed in the text format. Primarily, this means numbers, strings,
//! and vectors.
//!
//! The following number literals are supported:
//!
//! | Literal | Example |
//! |---------|---------|
//! | u32     | `10`    |
//! | u64     | `U10`   |
//! | s32     | `s10`   |
//! | s64     | `S10`   |
//! | u8      | `b5`    |
//! | f32     | `f10.5` |
//! | f64     | `F10.5` |
//! | v128    | `v10`   |
//! | s33     | `a10`   |
//!
//! TOOD: strings, vectors
//!
//! ### Aliases
//!
//! Almost everything else in dumbwasm is just an alias for a number that you'd have to write by
//! hand if you were to make a WebAssembly binary. For example, `i32` directly translates to the
//! number `0x7f`, no matter the context! You could even put `0x7f` in place of the magic value!
//!
//! Below is a table of such translations:
//!
//! | Syntax       | Output | Usage               |
//! |--------------|--------|---------------------|
//! | i32          | 0x7f   | Type                |
//! | i64          | 0x7e   | Type                |
//! | f32          | 0x7d   | Type                |
//! | f64          | 0x7c   | Type                |
//! | funcref      | 0x70   | Type                |
//! | externref    | 0x6f   | Type                |
//! | extern.func  | 0x00   | Import/export type  |
//! | extern.table | 0x01   | Import/export type  |
//! | extern.mem   | 0x02   | Import/export type  |
//! | form.func    | 0x60   | Type section form   |
//! | magic        | \0asm  | Beginning of module |
//! | version      | 1      | Beginning of module |
//! | true         | 0x01   | Generic             |
//! | false        | 0x00   | Generic             |

use crate::{lexer::Token, writer::Writer};
use logos::Logos;

mod lexer;
mod writer;

pub fn parse(input: &str) -> Vec<u8> {
    let lexer = Token::lexer(&input);
    let writer = Writer::new(lexer);
    writer.write()
}
