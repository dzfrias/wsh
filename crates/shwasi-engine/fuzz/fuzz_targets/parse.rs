#![no_main]

use libfuzzer_sys::fuzz_target;
use shwasi_engine::Parser;

fuzz_target!(|data: &[u8]| {
    let parser = Parser::new(data);
    let _ = parser.read_module();
});
