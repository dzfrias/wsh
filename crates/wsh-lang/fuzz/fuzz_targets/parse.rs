#![no_main]

use libfuzzer_sys::fuzz_target;
use wsh_lang::v2::{Parser, Source};

fuzz_target!(|data: &str| {
    let source = Source::new("<fuzz>", data.to_string());
    let _ = Parser::new(&source).parse();
});
