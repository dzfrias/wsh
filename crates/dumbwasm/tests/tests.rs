use dumbwasm::parse;
use std::fmt;

/// A struct who's Debug implementation prints the inner byte buffer in hexadecimal.
#[derive(PartialEq)]
struct PrettyBytes<'a> {
    buf: &'a [u8],
}

impl fmt::Debug for PrettyBytes<'_> {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        // format over multiple lines if there are more than 5 items
        let new_lines = self.buf.len() > 5;
        if new_lines {
            writeln!(f, "[")?;
        } else {
            write!(f, "[")?;
        }
        for (i, byte) in self.buf.iter().enumerate() {
            if new_lines {
                writeln!(f, "  {byte:#x},")?;
                continue;
            }
            if i == self.buf.len() - 1 {
                write!(f, "{byte:#x}")?;
                continue;
            }
            write!(f, "{byte:#x}, ")?;
        }
        write!(f, "]")?;
        Ok(())
    }
}

impl<'a> From<&'a [u8]> for PrettyBytes<'a> {
    fn from(buf: &'a [u8]) -> PrettyBytes<'a> {
        Self { buf }
    }
}

macro_rules! bytes {
    ($($byte:expr),*) => {
        {
            let bytes = PrettyBytes { buf: &[$($byte),*] };
            bytes
        }
    };
}

macro_rules! dumbwasm_test {
    ($name:ident, $input:expr, [$($byte:expr),*]) => {
        #[test]
        fn $name() {
            let buf = parse($input).unwrap();
            ::pretty_assertions::assert_eq!(bytes!($($byte),*), PrettyBytes::from(buf.as_slice()))
        }
    };
}

macro_rules! dumbwasm_test_failing {
    ($name:ident, $input:expr) => {
        #[test]
        fn $name() {
            use assert_cmd::Command;
            use std::{fs::File, io::Write};
            use tempdir::TempDir;

            let dir = TempDir::new(stringify!($name)).expect("could not create temp dir");
            let mut f = File::create(dir.path().join("input.dumbwasm"))
                .expect("could not create temp file");
            #[allow(clippy::string_lit_as_bytes)]
            f.write_all($input.as_bytes())
                .expect("could not write to temp file");
            let mut cmd = Command::cargo_bin(env!("CARGO_PKG_NAME")).unwrap();
            let assert = cmd.current_dir(dir.path()).arg("input.dumbwasm").assert();
            let out = assert.get_output();
            assert!(!out.status.success(), "command exited with exit code 0!");
            ::insta::assert_snapshot!(String::from_utf8_lossy(&out.stderr));
            dir.close().expect("could not close temp dir");
        }
    };
}

dumbwasm_test!(magic, "magic", [0x00, 0x61, 0x73, 0x6d]);
dumbwasm_test!(version, "version", [0x01, 0x00, 0x00, 0x00]);

dumbwasm_test!(
    aliases,
    "i32 i64 f32 f64 funcref externref extern.func extern.table extern.mem form.func true false",
    [0x7f, 0x7e, 0x7d, 0x7c, 0x70, 0x6f, 0x00, 0x01, 0x02, 0x60, 0x01, 0x00]
);
dumbwasm_test!(
    numbers,
    "4294967295 10.1",
    [0xff, 0xff, 0xff, 0xff, 0xf, 0x9a, 0x99, 0x21, 0x41]
);

dumbwasm_test!(
    comments,
    ";; comment
    (annotation)",
    []
);

dumbwasm_test!(
    type_indicators,
    "10<u8> 18446744073709551615<u64> 20.3<f64>",
    [
        0x0a, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0x1, 0xcd, 0xcc, 0xcc, 0xcc,
        0xcc, 0x4c, 0x34, 0x40
    ]
);
dumbwasm_test_failing!(bad_instruction, "i32.cons 10");

dumbwasm_test!(instructions, "i32.const 10 unreachable", [0x41, 0x0a, 0x00]);
dumbwasm_test!(
    instructions_with_prefix,
    "memory.init 0 0x00<u8>",
    [0xfc, 0x08, 0x00, 0x00]
);

dumbwasm_test_failing!(unknown_char, "i32 i32 ]");
dumbwasm_test_failing!(unclosed_array, "i32 [0, 1");
dumbwasm_test_failing!(unclosed_section, "function {  ");
dumbwasm_test_failing!(unclosed_instrs, "instrs {  ");
dumbwasm_test_failing!(out_of_place_type_indicator_float, "23<f32>");
dumbwasm_test_failing!(out_of_place_type_indicator_int, "10.1<u8>");

dumbwasm_test!(vectors, "[0, 1, 2, 3]", [0x04, 0x00, 0x01, 0x02, 0x03]);
dumbwasm_test!(
    type_section,
    "type { form.func [i32] [i32] }",
    [0x01, 0x05, 0x60, 0x01, 0x7f, 0x01, 0x7f]
);
dumbwasm_test!(
    instrs,
    "instrs { i32.const 10 i32.const 10 i32.add return }",
    [0x06, 0x41, 0x0a, 0x41, 0x0a, 0x6a, 0x0f]
);
dumbwasm_test!(
    strings,
    "\"Hello, world!\"",
    [0x0d, 0x48, 0x65, 0x6c, 0x6c, 0x6f, 0x2c, 0x20, 0x77, 0x6f, 0x72, 0x6c, 0x64, 0x21]
);

dumbwasm_test_failing!(suggests_general_keywords, "inst");

dumbwasm_test!(call_indirect, "call_indirect", [0x11]);
