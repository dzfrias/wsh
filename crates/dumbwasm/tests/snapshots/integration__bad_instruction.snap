---
source: crates/dumbwasm/tests/tests.rs
expression: "String::from_utf8_lossy(&out.stderr)"
---
error: unknown instruction: `i32.cons`

1|  i32.cons 10
 â€¢  ^^^^^^^^

help: perhaps you meant: `i32.const`

